### ISSUE TITLE
Live2d motion time reset on every show image

### ISSUE THREAD

**Issue by ParadoxalDeviant** [renpy/renpy#4731]
Hi  o/
Sorry to bother again about Live2d potential issues, but I think there might be something that was missed during the [last changes](https://github.com/renpy/renpy/commit/007aa1929128433e91ebf80ee8485d989281ffea) made for RenPy 8.1.1.

The last fixes are great and are working fine, but they created an issue where the motions are played from the beginning each time you call the image again. Here's an example to help describe what I mean:

```py
image epsilon walk = Live2D("anims/epsilon_walk",
                             top=.001,
                             base=.999,
                             loop=True,
                             seamless=True,
                             aliases={'smile': 'exp_smile', 'laugh': 'exp_laugh'}
                             )

label start:
    ""
    show epsilon walk move smile        # No issue, everything is fine
    ""
    show epsilon walk move laugh        # Correctly change expression, replay move.motion3 from the beginning
    ""
    show epsilon walk move smile        # Same as previous line
    ""
    return
```

I'll take a random guess on what's happening in `live2d.py` file that could lead to fixing the issue:
If `_sustain` attribute is not added to the attributes tuple returned by `_choosen_attribute`:
- `sustain` variable in [`_duplicate` method](https://github.com/renpy/renpy/blob/a33f891d65d3fd2b59a533c365800a59f8e15f0c/renpy/gl2/live2d.py#LL586C3-L586C3) will not switch to True
- `state.new_base_time` will not be equal to `state.old_base_time` in [`visit` method](https://github.com/renpy/renpy/blob/a33f891d65d3fd2b59a533c365800a59f8e15f0c/renpy/gl2/live2d.py#L445), which leads to having the motions going back to the beginning.

It's highly probable that it's not what's happening. I didn't dig deep enough to know what I'm talking about, but having `state.new_base_time` always equal to `state.old_base_time` seems to fix the issue _(and create another issue by not resetting the motion time when you switch to a new motion)_.

Hope someone can find a solution about it.

**TLDR: Current motion is played from the beginning when you call the image after you already displayed it (see above example code)**

**Comment by renpytom** [comment]
I don't think this is an error. When you're specifying show epsilon walk move laugh, you're asking Ren'Py to restart the motion. If you didn't want to restart the motion, you'd have written "show epsilon laugh", and let the motion continue.

**Comment by ParadoxalDeviant** [comment]
Alright! Bad habit to always precise the motion that should be played, even though it's already playing. It wasn't a problem before so I didn't even think about changing this part of my code.
Though I have some concerns about attribute filters, at least the way I use them...

I changed a bit a filter I used, to match the way motions are working right now:
```py
init -5 python:
    def cg_epsilon_walk_filter(attrs):

        # List of all available motions and nonexclusive expressions this image has
        motions = ["move"]
        exps = ["smile", "laugh"]

        # Store all motions and nonexclusive expressions passed through attributes
        attr_motions = attr_exps = []
        for attr in attrs:
            if attr in motions:
                attr_motions.append(attr)
            if attr in exps:
                attr_exps.append(attr)

        # Select the first expression that was passed in attributes as the only expression to use.
            # Otherwise select the first expression in the list of all expression available
        exp = attr_exps[0] if attr_exps else exps[0]
#        motion = attr_motions[0] if attr_motions else motions[0]

#        return tuple([motion, exp])

        # Return the defined expressions and the passed motion if there was one, otherwise just return the expression
        if attr_motions:
            return tuple([attr_motions[0], exp])
        else:
            return tuple([exp])



image epsilon walk = Live2D("anims/epsilon_walk",
                             top=.001,
                             base=.999,
                             loop=True,
                             seamless=True,
                             attribute_filter=cg_epsilon_walk_filter,
                             #aliases={'smile': 'exp_smile', 'laugh': 'exp_laugh'}     <- I removed 'exp_' from the filenames
                             )

label start:
    ""
    show epsilon walk move        # Enter attribute_filter with attrs = ('move',) and exit it with attrs = ('move', 'smile')
    ""
    show epsilon walk laugh        # Enter attribute_filter with attrs = ('laugh', 'smile',) and exit it with attrs = ('laugh')
    return
```

In this case I use my attribute filter to have easy switch between `smile` and `laugh` expressions. Because there must be one of them displayed at all time, if I don't tell in the `show` statement which one I want, it takes `smile` by default... but I can't do the same with `move`...
It's seems like impossible to reduce the first line to `show epsilon walk`. If I change my filter to automatically add `move` to the attributes I will not know it was added previously because, contrary to the expressions, the motion attribute of the image does not stay in the attributes tuple when the second `show` is called.

Is there a way to improve this filter to automatically add a default motion on the first `show` of the image without having it to reset on the second `show`? Is there some variable that I don't know that can tell me when I enter the filter if a motion is currently in play and which one it is (in case I can use multiple motions)?

**Comment by renpytom** [comment]
It looks like your function hit on a weird edge case, where the change in 8.1.1 was to move the test for sustain where attribute filter is applied. 

I think using an attribute filter here is just wrong - you're doing this at the wrong level. The function you want is config.default_attribute_callbacks, which happens on a higher level, where the attributes that are part of the first show are available to it. 

You'd write:
```renpy
init python:
     def add_smile(name):
          if "laugh" not in name and "smile" not in name:
                return ("smile",)
          else:
                return ()

default config.default_attribute_callbacks['epsilon'] = add_smile
```

Check out the documentation here: 

https://www.renpy.org/doc/html/config.html#var-config.default_attribute_callbacks

You're right that attribute_filter doesn't have enough information, but that's because determining default attributes is best done at a higher level.

**Comment by ParadoxalDeviant** [comment]
You are right. I was trying to make things work based on a loophole I accidentally found. I just didn't realized it wasn't suppose to work like this until the recent changes.

I was aware of `config.adjust_attributes` but not `config.default_attribute_callbacks` and doing things at a higher level seems a better solution. I'll try out some things based on how `config.default_attribute_callbacks` works and surely I'll be able to fix my broken code.

We can close this issue, thanks for everything  &nbsp; &nbsp;o/
\
\
\
_**PS**: It could be great to have a method returning if yes or no your Live2d image is currently playing a motion and which one it is._

**Comment by renpytom** [comment]
I don't think there's further action needed here. I see the ps, but I'll need to make more changes in Ren'Py before it's possible.
